File: solutions/228-summary-ranges/solution.py
from typing import List


class Solution:
    def summary_ranges(self, nums: List[int]) -> List[str]:
        result: list[str] = []
        i, n = 0, len(nums)

        while i < n:
            start = nums[i]
            while i + 1 < n and nums[i + 1] == nums[i] + 1:
                i += 1
            end = nums[i]
            if start == end:
                result.append(str(start))
            else:
                result.append(f"{start}->{end}")
            i += 1

        return result


if __name__ == "__main__":
    tests = [
        ([0, 1, 2, 4, 5, 7], ["0->2", "4->5", "7"]),
        ([0, 2, 3, 4, 6, 8, 9], ["0", "2->4", "6", "8->9"]),
        ([], []),
        ([1], ["1"]),
    ]
    sol = Solution()
    for nums, expected in tests:
        result = sol.summary_ranges(nums)
        status = "PASS" if result == expected else "FAIL"
        print(f"{status} | nums={nums!r} => {result}")

File: solutions/228-summary-ranges/solution.go
package main

import (
	"fmt"
	"strconv"
)

func summaryRanges(nums []int) []string {
	result := make([]string, 0)
	i, n := 0, len(nums)

	for i < n {
		start := nums[i]
		for i+1 < n && nums[i+1] == nums[i]+1 {
			i++
		}
		end := nums[i]
		if start == end {
			result = append(result, strconv.Itoa(start))
		} else {
			result = append(result, fmt.Sprintf("%d->%d", start, end))
		}
		i++
	}
	return result
}

func equal(a, b []string) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

func main() {
	tests := []struct {
		nums     []int
		expected []string
	}{
		{[]int{0, 1, 2, 4, 5, 7}, []string{"0->2", "4->5", "7"}},
		{[]int{0, 2, 3, 4, 6, 8, 9}, []string{"0", "2->4", "6", "8->9"}},
		{[]int{}, []string{}},
		{[]int{1}, []string{"1"}},
	}

	for _, tc := range tests {
		result := summaryRanges(tc.nums)
		status := "PASS"
		if !equal(result, tc.expected) {
			status = "FAIL"
		}
		fmt.Printf("%s | nums=%v => %v\n", status, tc.nums, result)
	}
}
